src/main/java/org/terasology/flyingIslands/FlyingIslandDensityProvider.java

```diff
@@ -32,28 +33,31 @@ public void process(GeneratingRegion region) {
         DensityFacet densityFacet = region.getRegionFacet(DensityFacet.class);
         SurfacesFacet surfacesFacet = region.getRegionFacet(SurfacesFacet.class);
 
-        for (Map.Entry<BaseVector3i, FlyingIsland> entry : flyingIslandFacet.getWorldEntries().entrySet()) {
+        for (Map.Entry<Vector3ic, FlyingIsland> entry : flyingIslandFacet.getWorldEntries().entrySet()) {
 
             Vector3i basePosition = new Vector3i(entry.getKey());
             FlyingIsland flyingIsland = entry.getValue();
 
             int extent = (int) flyingIsland.getOuterRadius();
-            int top = FlyingIsland.MAXHEIGHT;
+            int maxTop = FlyingIsland.MAX_DEPTH + FlyingIsland.MAX_HEIGHT;
 
             for (int i = -extent; i <= extent; i++) {
                 for (int k = -extent; k <= extent; k++) {
-                    Vector3i position = new Vector3i(i, top, k).add(basePosition);
+                    Vector3i position = new Vector3i(i, maxTop, k).add(basePosition);
 
-                    int height = flyingIsland.getHeightAndIsLava(position.x, position.z);
+                    int depth = flyingIsland.getDepth(position.x, position.z);
+                    int top = flyingIsland.getTop(position.x, position.z);
 
-                    if (height > 0 && surfacesFacet.getWorldRegion().encompasses(position)) {
-                        surfacesFacet.setWorld(JomlUtil.from(position), true);
+                    Vector3i surface = position.add(0, -top, 0);
+
+                    if (depth > 0 && surfacesFacet.getWorldRegion().contains(position)) {
+                        surfacesFacet.setWorld(surface, true);
                     }
 
-                    for (int j = top; j > top - height; j--) {
+                    for (int j = maxTop - top; j > maxTop - depth - top; j--) {
                         Vector3i position2 = new Vector3i(i, j, k).add(basePosition);
-                        if (densityFacet.getWorldRegion().encompasses(position2)) {
-                            densityFacet.setWorld(position2, top - j + 1);
+                        if (densityFacet.getWorldRegion().contains(position2)) {
+                            densityFacet.setWorld(position2, maxTop - j + 1);

```

Comment:
This should be the distance from the top, rather than the distance from the max top. Otherwise, areas where the surface of the flying island is lower will have thinner dirt or sand.

Reply:
actually i renamed what earlier was top -> maxTop since it is the maximum possible top height of an island and changing it to top seems to prevent any islands from being formed

Reply:
This is still incorrect. The density at the surface block should be 1, and the height of the surface (relative to `basePosition`) is `maxYSize-height`, not `maxYSize`.